### ISSUE TITLE
Command line options not working properly

### ISSUE THREAD

**Issue by gonzaponte** [next-exp/IC#307]
Some command line options are not working properly. For example:

- -n, --nevents. It does not work because the variable used in cities is named nmax,
- --runall. This one is simply ignored.
- -s, --skip. This one is also ignored.
- -f, --first-event. This one is also ignored, and should be the same as -s.

I have only checked the -n option, but I didn't find anything about the others in the code.

**Comment by jacg** [comment]
We need automated tests for these.

There are two options:

+ execute a command line and parse the output (ugly, flaky, but less likely to bypass something important in the chain)

+ feed a fake `argv` directly into a python function, and check the return value (much cleaner, but it might miss some step which could give rise to errors)

**Comment by jjgomezcadenas** [comment]
I have assigned fixing this and automated tests to @gonzaponte and @jacg OK?

**Comment by gonzaponte** [comment]
ok

**Comment by jacg** [comment]
> -n, --nevents. It does not work because the variable used in cities is named nmax,

Which name do we prefer `nmax` or `nevents`?

**Comment by jacg** [comment]
> -s, --skip. This one is also ignored.
> -f, --first-event. This one is also ignored, and should be the same as -s.

Let's pick one of them. Which one?

**Comment by jjgomezcadenas** [comment]
-nmax 
-f

**Comment by jacg** [comment]
Just like I prefer `first-event` over `skip` because it contains 'event' and therefore self-documents what it is talking about, I would prefer `nevents` over `nmax`.

**Comment by gonzaponte** [comment]
I suggest `n-events-max` and `first-event` to make everybody happy.

**Comment by jacg** [comment]
> `n-events-max`

That's what I was going to suggest.

**Comment by jacg** [comment]
What is the relationship between `--run-all` and `--n-events-max` (née `--nevents`) supposed to be?

**Comment by gonzaponte** [comment]
I was just writing about that. `--run-all` is supposed to override `--n-events-max` so all events are processed. But since we can put `-n -1`, I think it is redundant.

**Comment by jacg** [comment]
Given that `--run-all` and `--n-events-max` (and their corresponding config file entries) can give rise to contradictory information, I think that we should do one of the following:

+ Raise an exception if these two settings contradict eachother
+ Eliminate `--run-all` altogether, and make an absence of `--n-events-max` imply 'run all'.

I am strongly in favour of the latter.

**Comment by gonzaponte** [comment]
> I am strongly in favour of the latter.

Me too.

**Comment by jjgomezcadenas** [comment]
Me2

**Comment by jacg** [comment]
Of course, the absence of `--n-events-max` on the command line (or the absence of `n_events_max` in a config file) should *not* override a setting of the option lower down in the configuration implicitly, but an *explicit* override should be possible.

I see a number of ways of doing it:

1. use `-1`  (or any negative number) to mean 'run_all'
2. use `0` to mean 'run_all'
3. use `None` to mean run_all'
4. use generalized booleans: *any* Python false-like value means 'run_all'
5. introduce a `Maybe` or `Option` type (à la Haskell / ML / Scala / Rust ... hell, even Java 8 and C++ 17 have them!) and use that.

Any preferences? (I'll withhold mine for now, so as not to influence you too much.)

**Comment by gonzaponte** [comment]
> Any preferences? (I'll withhold mine for now, so as not to influence you too much.

Are you saying that you do not prefer option 5 as the best option EVER? ;)

Some comments:
- Option 1 is very old-fashioned. I think *art* uses it, @jjgomezcadenas will then say "HELL NO", however, I am comfortable using it.
- Options 2-4 seem a bit odd... What does run "run 0/None/False" events mean? You may say "What does run -1 events mean?"... But it seems more natural to me.
- Can you expand on option 5? I am a bit familiar with `Maybe` and `Option`, but I would like some detail about the implementation. Or simply and better, push an example!

**Comment by jacg** [comment]
Of course option 5 is the best ever ... exept for this line in the Zen of Python

> [...] practicality beats purity.

The absence of such a type from the standard library, or even PyPI, would require us to create such a type, and it doesn't seem a very practical approach in this context.

> What does run "run 0/None/False" events mean?

It seems quite weird, doesn't it? But if you take into account that we have renamed this option as `--n-events-max`, the question turns into 'what is the maximum number of events we want to run?'. Then the answer `None` means that there is no maximum, and that we should run all of them.

Perhaps it would be even better if we named it `--n-events-limit`.

At which point it should be pretty clear which my favourite is.

**Comment by jacg** [comment]
... as for option 5, while I'd love to play with an implementation of it, I really don't have time for that right now.

**Comment by gonzaponte** [comment]
> The absence of such a type from the standard library, or even PyPI

I guess you have already checked [PyMonad](https://pypi.python.org/pypi/PyMonad/).

**Comment by jacg** [comment]
I have, though it's been a long time since I took a close look at it. There's too much stuff there and I don't think we want to let it anywhere near IC ... just yet ... coz once you let the genie out of the bottle ...

Actually, I'm not sure when @jjgomezcadenas was planning on telling you that for the next major version IC will be rewritten in Haskell. When he does, you'll realize that we won't need this library, as all these monads will be in the standard library.

**Comment by gonzaponte** [comment]
> Actually, I'm not sure when @jjgomezcadenas was planning on telling you that for the next major version IC will be rewritten in Haskell.

He has been leaking something here and there... I will resume my Haskell learning soon then...

**Comment by jacg** [comment]
After struggling with the implementation a bit (`None` has all sorts of other meanings deep within our handling of the configuration) I propose another approach:

rather than `--n-events-max None` on the CLI and `n_events_max = None` in the config files, I propose that we have something like

`--n-events-max All` and `n_event_max = All` in order to explicitly override any limits set lower down in the configuration.

Opinions?

**Comment by gonzaponte** [comment]
`All` as a variable or as a string?

**Comment by jacg** [comment]
As a singleton *object* whose type is not `str` and whose sole purpose is to identify that situation. A Lisp symbol, if you like.

**Comment by jacg** [comment]
The first-order implementation of `All` (or whatever we would call it) might be

    class All:
        pass

We would probably want to endow it with a helpful docstring.

**Comment by jacg** [comment]
I prefer the class approach to the perhaps more obvious

    All = object()

because

1. The class has more informative `__repr__` and `__str__`.
2. There's somewhere to put a docstring.

Alternatively, we could write a symbol type, but I'm reluctant to do that until we have more than one symbol.

**Comment by gonzaponte** [comment]
Conceptually, we just want a more obvious version of `n_events_max = -1`, right? So we could simply do `All = -1`. Another option is to implement the comparison operators for an empty class `All` so it does never stop the loop.

We would still have to handle the "All" (string) coming from the CLI, wouldn't we?

**Comment by jacg** [comment]
> Conceptually, we just want a more obvious version of `n_events_max = -1`, right?

Not really. As you pointed out yourself, semantically it doesn't make sense.

> So we could simply do `All = -1`.

Urgh! I know that a lot of Python libraries do that sort of thing, but that's because they were written by people who don't know better: people who don't realize how useful symbols are (and have probably never thought about the concept). The same goes for those who use strings instead of symbols (e.g. matplotlib), although strings are better than integers for this sort of purpose.

There are at least two big problems with `All = -1` (and and bazillion smaller ones):

1. It's `__repr__` and `__str__` will both give you `-1` which really doesn't help debugging.
2. It will be indistinguishable from any other integer, while its meaning is completely different.
3. You will be able to do arithmetic on it.
4. There is no indication what the set of meaningful choices is, so you will be able to use `-2` (or any other integer instead) and really confuse the reader (or introduce a bug).
5. etc.

Fundamentally, an integer is the wrong type for this sort of thing. So are strings. The whole purpose of `All` is to have a unique identity and convey the meaning we want. You shouldn't be able to perform any other operations on it besides comparing it to other things.

In summary, `All` should be a symbol.

Aaah, but hang on, enumerations landed in Python 3.4 and they are a decent approximation to symbols. (I remember being overjoyed at this happening at the time, and then I forgot about it: brain like a sieve!)

**Comment by jacg** [comment]
> We would still have to handle the "All" (string) coming from the CLI, wouldn't we?

Yup. No problem.

**Comment by gonzaponte** [comment]
Sorry, I didn't mean to *actually* use `All = -1`, it was rather a picture in my mind. I agree that `All` should be a symbol, or something very similar and that, for that purpose, enumerations sound good.

**Comment by jacg** [comment]
Until we have more than one of symbol relating to the same concept, I think that `Enum`s are more trouble than they're worth. I'll go with `class All: pass` for now.

If and when we need anything more sophisticated, let's try the `enum` module.

**Comment by jacg** [comment]
Summarizing the state of play:

+ `--skip` which was a synonym for `--first-event`, has been removed.
+ `--nevents` (known as `nmax` in the code) has been renamed `--n-events-max`.
+ `--run-all` has been removed and replaced with `--n-events-max All`.

**Comment by jacg** [comment]
`--print-mod` is used in the CLI, but `nprint` is used in the config files. Which do we want to keep? Or do we want to change it to something else?

**Comment by jacg** [comment]
For testing purposes I need a combination of city + test data file which will process all events in the file as quickly as possible, but there should be a number of events (not just a single one) in the file.

Which city + data file is best suited for this?

**Comment by jjgomezcadenas** [comment]
Sorry about being so quiet about this whole thing (may best for everybody),  very busy days just back in Spain.

Also, I see that the master has disclosed that IC2.0 will be written in Haskell. I am sure everybody in NEXT will be happy about it!

As for city + data. Probably best to use Penthesilea, which has full input output but light input (PMAPS) (while Irene, for example has RWF files which are much heavier).

@gonzaponte can you send @jacg a PMAP file?

Thanks!

**Comment by jacg** [comment]
`KrMC_pmaps.h5` should do the trick, shouldn't it? It contains 27 events, and they get processed at 0.002 s / event.

**Comment by jjgomezcadenas** [comment]
yes, I think so

**Comment by jacg** [comment]
How many events should be processed with CLI flags `--first-event 10 --n-events-max 15` (assuming that there's a bazillion events in the file)?

+ 6 (events from 10 to 15 inclusive)
+ 15 (15 events starting with event 10)
+ something else?

**Comment by jjgomezcadenas** [comment]
I assume first answer is right

**Comment by jacg** [comment]
If that is the case, could we please replace `--first-event` and `--n-events-max` with a single parameter called something like `--event-range`?

**Comment by jjgomezcadenas** [comment]
absolutely

**Comment by jacg** [comment]
In which case I await confirmation from @gonzaponte, @jmbenlloch, @abotas, and anyone else who cares, that this is indeed the interpretation we want.

If we go ahead with this, how about something like:

+ `--event-range first last`
+ `--event-range first 10`
+ `--event-range 20 last`

It might even be possible to make a shortcut for the first

+ `--event-range all` (but here it would be prudent to distinguish it from the builtin `all`)

What should the abbreviated single-letter version of `--event-range` be? Remember it would be replacing `--first-event` (`-f`), (with synonym `--skip` (`-s`)) and `--n-events-max` (`-n`) (née `--nmax` / `--nevents`).

The obvious candidates seem to be

+ `-r` : currently taken by `--run-number`
+ `-e` : currently unused.

**Comment by gonzaponte** [comment]
> --print-mod is used in the CLI, but nprint is used in the config files. Which do we want to keep? Or do we want to change it to something else?

`print-mod` seems more obvious

> `KrMC_pmaps.h5` should do the trick, shouldn't it? It contains 27 events, and they get processed at 0.002 s / event.

The data in that file are the PMAPS for events 23-51. Some of those events contain only S1, some of them contain only S2, and some of them contain both. I think it works well for your purposes.

> How many events should be processed with CLI flags --first-event 10 --n-events-max 15 (assuming that there's a bazillion events in the file)?

I would interpret it as "run 15 events starting at event 10".

> could we please replace --first-event and --n-events-max with a single parameter called something like --event-range?

> If we go ahead with this, how about something like:
>
> - `--event-range first last`
> - `--event-range first 10`
> - `--event-range 20 last`

Sound good to me.

> It might even be possible to make a shortcut for the first
> - `--event-range all` (but here it would be prudent to distinguish it from the builtin all)

Why `all` and not the symbol `All`?


> The obvious candidates seem to be
>
> - `-r` : currently taken by `--run-number`
> - `-e` : currently unused.

An abbreviation for `--run number` that is not `-r` sounds a bit unnatural to me, so I'd stick to the latter.

**Comment by jacg** [comment]
Do we agree that `--event-range 10 20` should give us 10 events? I.e. event-range should define a semi-open interval, just like Python's ranges and slices?

**Comment by jjgomezcadenas** [comment]
agree with @gonzaponte and also agree with semi open interval a la python, as suggested by @jacg

**Comment by jacg** [comment]
> Why `all` and not the symbol `All`?

It's a minor N-th order point: I'm just trying to be nice to the user and not force annoying capitalization in the middle of commands.

**Comment by jjgomezcadenas** [comment]
Seems indeed a nice thing to do.

**Comment by gonzaponte** [comment]
> Do we agree that --event-range 10 20 should give us 10 events? I.e. event-range should define a semi-open interval, just like Python's ranges and slices?

I do.

> It's a minor N-th order point: I'm just trying to be nice to the user and not force annoying capitalization in the middle of commands.

Ok.

**Comment by jacg** [comment]
OK, here's a second iteration concerning the `--event-range` interface:

+ `-e all` : all events in the files
+ `-e 34` : first 34 events in the files
+ `-e 10 15` : 5 events, starting at number 10
+ `-e 12 last` : skip the first 12 events (I trust that event numbers start at `0`)

How does that look?

I welcome suggestions for a short synonym for `all` that does not clash with a Python builtin.

**Comment by gonzaponte** [comment]
Looks good to me.

> I welcome suggestions for a short synonym for all that does not clash with a Python builtin.

full?

**Comment by jacg** [comment]
`full` fits well with `--event-range`. It bothers me a bit with `-e` because I guess I was thinking of `-e` as 'events'. Which raises the question: do we prefer `--event-range` or `--events`?

**Comment by gonzaponte** [comment]
`--event-range` fits better, I think. But then the case `--event-range X` to run the first X files can be misleading.

**Comment by jacg** [comment]
>  `--event-range X` to run the first X files can be misleading.

Not to anyone remotely familiar with Python, I would hope.

The cost of using `all` rather than `full` isn't all that great: it's just that you'll be in for a surprise if you ever try to use the builtin `all` in a config file. Which is pretty unlikely, I hope.

If we use `all` rather than `full`, that shifts the balance a bit towards `--events`.

**Comment by jacg** [comment]
After a major rewrite resulting from the move from `--first-event` and `--n_events_max` to `--event_range` I am getting a few test failures. Three of these are essentially

    assert nrequired == nactual

and all of these fail because `nrequired == 2` while `nactual == 1`. This happens in

+ `test_irene_electrons_40keV`
+ `test_isidora_electrons_40keV`
+ `test_irene_runinfo_run_2983`

As I am not intimately familiar with these tests or the input files that they use, can anyone help me understand why the expected number of events should be 2?

I can push a messy commit to this PR, if you want to have a look at the new version, but it's still very much WIP.

**Comment by gonzaponte** [comment]
I can't see anything why these tests would fail. The messy commit would help.

**Comment by jacg** [comment]
> The messy commit would help.

It's there at the end of #324. Which is where meant to ask my question, sorry.

**Comment by gonzaponte** [comment]
The tests are still using `n_events_max`. Am I missing something?
